device/alpaka/src/clusterization/measurement_sorting_algorithm.cpp

```diff
@@ -0,0 +1,46 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Library include(s).
+#include "traccc/alpaka/clusterization/measurement_sorting_algorithm.hpp"
+
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+#include <thrust/sort.h>
+#endif
+
+namespace traccc::alpaka {
+
+measurement_sorting_algorithm::measurement_sorting_algorithm(vecmem::copy& copy)
+    : m_copy{copy} {}
+
+measurement_sorting_algorithm::output_type
+measurement_sorting_algorithm::operator()(
+    const measurement_collection_types::view& measurements_view) const {
+
+    // Get the number of measurements. This is necessary because the input
+    // container may not be fixed sized. And we can't give invalid pointers /
+    // iterators to Thrust.
+    const measurement_collection_types::view::size_type n_measurements =
+        m_copy.get().get_size(measurements_view);
+
+    // Sort the measurements in place
+    // TODO: This obviously won't work for HIP, but we don't have a HIP
+    // equivalent of Thrust yet.
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+    thrust::sort(thrust::device, measurements_view.ptr(),
+                 measurements_view.ptr() + n_measurements,
+                 measurement_sort_comp());
+#elif !defined(ALPAKA_ACC_GPU_HIP_ENABLED)
+    std::sort(measurements_view.ptr(), measurements_view.ptr() + n_measurements,
+              measurement_sort_comp());
+#endif
```

Comment:
This won't work for HIP / isn't very ideal.

@StewMH is going to be looking into getting `rocThrust` support, in which case this (and the later uses over in the track finding) should be much easier. Alpaka does have some thrust-like bits around this in https://github.com/alpaka-group/vikunja (transforms and reduces as far as I see), but not a direct replacement for a sort.

My understanding of PR #550 by @krasznaa was that this is basically a necessary evil, and eventually something like an algorithm re-design may mean it isn't needed, but for now I think this is "good enough".

Reply:
Should be fixed with the recent rocThrust change, and I'm testing it more thoroughly with my current work.